xdsl/dialects/qref.py

```diff
@@ -70,6 +94,17 @@ def print(self, printer: Printer):
 class HGateOp(QRefBase):
     name = "qref.h"
 
+    def ssa_op(self) -> qssa.HGateOp:
```

Comment:
Best to keep dialect conversion functions in the converter rather than on the dialect.

Reply:
I am unsure if this is desirable, as I don't see a way around having a big `if isinstance` chain in the converter without doing it this way. Another advantage of this method is that if a new operation is added, then the implementation of these two functions is forced

Reply:
You could match on the specific op, create a new one, and move any shared code to a helper function., e.g. something along the lines of:

```
class ConvertHGateToQssaPattern(RewritePattern):
    @op_type_rewrite_pattern
    def match_and_rewrite(self, op: qref.HGateOp, rewriter: PatternRewriter):
        replace_with_ssa(
            op,
            qssa.HGateOp.create(
                operands=self.operands,
                result_types=(qssa.qubit,),
                attributes=self.attributes,
            ),
            rewriter,
        )
```

Reply:
This still runs into the problem of not forcing the conversion to be written for new ops, though I guess a test could be written that ensures this

Reply:
I kind of like the current setup. The only thing I might suggest is to have a `QuantumGate` trait instead of `is_gate()` but even that seems like a nit, and like less good a way to guarantee that each op will have to implement these methods when it's added. It also feels like the two dialects are coupled in a pretty fundamental way, which seems like an ok design decision, making the abstract method style of conversion quite natural. I'm not sure in what situation in the future we would regret merging this PR as-is.

Reply:
> This still runs into the problem of not forcing the conversion to be written for new ops, though I guess a test could be written that ensures this

This could also be a feature. Forcing a conversion to be made also implies an unwritten guarantee that there will always be a valid 1:1 conversion, and that neither of these dialects will ever hold concepts that are not directly convertible, now or in the future. I just wanted to draw attention to the fact that this tight coupling might be a rather committal decision. I don't fully know the plans for these dialects, so if you're happy then please go ahead and merge.